tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,300 @@
+    }
+
+    fn set_to_initialized(&self) {
+        self.value_set.store(true, Ordering::Release);
+    }
+
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    unsafe fn set_value(&self, value: T) {
+        self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+    }
+
+    /// Tries to get a reference to the value of the OnceCell.
+    ///
+    /// Returns [`OnceCellError::NotInitialized`] if the value of the OnceCell
+    /// hasn't previously been initialized
+    ///
+    /// [`OnceCellError::NotInitialized`]: crate::sync::OnceCellError::NotInitialized
+    pub fn get(&self) -> Result<&T, OnceCellError> {
+        if self.initialized() {
+            Ok(unsafe { self.get_unchecked() })
+        } else {
+            Err(OnceCellError::NotInitialized)
+        }
+    }
+
+    /// Sets the value of the OnceCell to the argument value.
+    ///
+    /// If the value of the OnceCell was already set prior to this call, then
+    /// [`OnceCellError::AlreadyInitialized`] is returned
+    ///
+    /// [`OnceCellError::AlreadyInitialized`]: crate::sync::OnceCellError::AlreadyInitialized
+    pub fn set(&self, value: T) -> Result<(), OnceCellError> {
+        if !self.initialized() {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or another thread has intialized the value and
+            // closed the semaphore, in which case self.initialized is true and we
+            // don't set the value
+            match self.sema.try_acquire() {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        self.set_to_initialized();
+                        self.sema.close();
+                        return Ok(());
+                    } else {
+                        panic!("acquired the permit after OnceCell value was already initialized");
+                    }
+                }
+                _ => {
+                    if !self.initialized() {
+                        panic!(
+                            "couldn't acquire a permit even though OnceCell value is uninitialized"
+                        );
+                    }
+                }
+            }
+        }
+
+        Err(OnceCellError::AlreadyInitialized)
+    }
+
+    /// Tries to initialize the value of the OnceCell using the async function `f`.
+    /// If the value of the OnceCell had already been initialized prior to this call
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F>(&self, f: F) -> &T
+    where
+        F: Fn() -> Pin<Box<dyn Future<Output = T> + Send>>,
```

Comment:
Can be simplified to
```suggestion
    pub async fn get_or_init<F, Fut>(&self, f: F) -> &T
    where
        F: FnOnce() -> Fut,
        Fut: Future<Output = T>,
```